<!DOCTYPE html>
<html lang="en">
	<head>
		<meta charset="utf-8">
		<title>vcf.iobio.io</title>

        <script>
          (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
          (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
          m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
          })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

          ga('create', 'UA-47481907-3', 'auto');
          ga('send', 'pageview');

        </script>


        <link href='http://fonts.googleapis.com/css?family=Overlock+SC' rel='stylesheet' type='text/css'>
        <link href='assets/css/quicksand.css' rel='stylesheet' type='text/css'>
        <link href="assets/css/flat-ui.css" rel="stylesheet" type='text/css'>
        
        <link href="assets/css/selectize.css" rel="stylesheet" type='text/css'>
        <link href="assets/js/bootstrap/css/bootstrap.css" rel="stylesheet" type='text/css'>
        <!-- <link href="assets/css/bootstrap.css" rel="stylesheet" type='text/css'> -->
		<link href='components/nprogress/nprogress.css' rel='stylesheet' type='text/css'>
        <link rel="stylesheet" href="components/alertify.js/themes/alertify.core.css" />
        <link rel="stylesheet" href="components/alertify.js/themes/alertify.default.css" />
        <link href="assets/css/iobio.css" rel="stylesheet" type='text/css'>
		<link href="assets/css/iobio.viz.min.css" rel="stylesheet" type='text/css'>
		<link href="assets/css/vcf.iobio.css" rel="stylesheet" type='text/css'/>

		<script type="text/javascript" src="components/d3/d3.js"></script>
        <script type="text/javascript" src="components/jquery/dist/jquery.min.js"></script>
        <script type="text/javascript" src="components/nprogress/nprogress.js"></script>
        <script type="text/javascript" src="components/alertify.js/lib/alertify.min.js"></script>
        <!-- <script type="text/javascript" src="assets/js/bootstrap.min.js"></script> -->
        <script type="text/javascript" src="assets/js/bootstrap/js/bootstrap.min.js"></script>
        <script type="text/javascript" src="assets/js/class.js"></script>
        <script type="text/javascript" src="assets/js/rdp.js"></script>
        <script type="text/javascript" src="assets/js/selectize.js"></script> 
        <script type="text/javascript" src="assets/js/iobio.viz.min.js"></script>

        <script src="assets/js/binary.js"></script>

        <script src="assets/js/jdataview.js"></script>
        <script src="assets/js/jparser.js"></script>
        <script src="assets/js/inflate.js"></script>
        <script src="assets/js/pako_deflate.min.js"></script>
        <script src="assets/js/jsbgzf.js"></script>
        <script src="assets/js/js-bv-sampling.js"></script>
        <script src="assets/js/js-local-vcf.js"></script> 

           <!--  <script type="text/javascript" src="assets/js/binary-vcf.js"></script> -->
        <script type="text/javascript" src="assets/js/iobio.min.js"></script>

        <script>
          $(document).ready( function(){
            // Help modals
              $("#density_help").click(function() {
                $("#densityModal").modal({
                  show: true
                });
              });
              $("#tstv_help").click(function() {
                $("#tstvModal").modal({
                  show: true
                });
              });
              $("#af_help").click(function() {
                $("#afModal").modal({
                  show: true
                });
              });
              $("#bases_help").click(function() {
                $("#basesModal").modal({
                  show: true
                });
              });
              $("#variant_help").click(function() {
                $("#variantModal").modal({
                  show: true
                });
              });
              $("#indel_help").click(function() {
                $("#indelModal").modal({
                  show: true
                });
              });
              $("#quality_help").click(function() {
                $("#qualityModal").modal({
                  show: true
                });
              });
            });
          </script>
        
	</head>
	<body>
        <header>
          
        </header>

        
        <div id="selectData">
            <header>
                <a href="">vcf<span style="color:rgb(200,200,200)">.iobio.io</span></a>
            <div style="float:right" class="iobio-project" style="padding-top:20px">
                <a href="http://iobio.io" style="margin: 0px 5px 0px 5px;">an iobio project</a>                     
            </div>
            
            </header>    
               
            <div style="width:700px; margin-left:auto; margin-right:auto; margin-top: 70px">

                <div style="margin-left:auto;margin-right:auto;font-size: 28px; color: rgb(110,110,110); margin-bottom:40px">                
                    <center>examine your variant file in seconds</center>
                </div>
                
                <div class="file" onClick="onFileButtonClicked()">
                    <input type="file" name="files[]" id="file"  multiple />
                    <label class="file-button" for="file" >choose vcf file</label>
                </div>
                <div class="file-button" style="text-align:center" onClick="displayVcfUrlBox()">choose vcf url
                </div>
                <div style="clear:both"></div>

            </div>
            <div id="vcf-url" style="visibility:hidden;">
                <div class="arrow_box" style="display:inline-block;width:670px;">
                    <input id="url-input" value="http://s3.amazonaws.com/vcf.files/ExAC.r0.2.sites.vep.vcf.gz" placeholder="vcf.gz URL" ></input>
                    <input id="url-tbi-input" value="" placeholder="tbi URL (optional)"></input>
                </div>
            </div>
            <div id="file-alert" class="alert alert-danger hide" role="alert">
            </div>


            <div id="go-panel" class="hide">
              <div id="select-species-box" style="font-size:18px;width:230px;float:left;">
                <span id="species-label">Species</span>
                <select id="select-species" class="selectized" placeholder="select species...">
                </select>
              </div>
              <div id="select-build-box" style="font-size:18x;width:270px;float:left;margin-left:20px;">
                <span id="build-label">Genome Build</span>
                <select id="select-build" class="selectized" placeholder="select genome build...">
                </select>
              </div>
            
              <button id="url-go-button" class="go-button btn disabled" onclick="loadFromUrl()">Go</button>
              <button id="file-go-button" class="hide go-button btn disabled" onclick="loadFromFile()">Go </button>

              <div id="species-build-warning" class="alert alert-danger hide" style="float:left;width:100%">                
              </div>              
            </div>

            <div style="clear:both"></div>
            <div id="info">                                
                <ul>
                        <li><a href="help.html">File Requirements</a></li>
                        <li><a href="http://bam.iobio.io/license.html">License</a></li>
                        
                </ul>           
            </div>

            
            <div style="width:700px; margin-left:auto; margin-right:auto; margin-top: 10px">

                <strong>Variant files to try</strong>

                <table cellspacing="5px" width="700px">

                <tr>
                    <td>
                        Exome Aggregation Consortium (ExAC)
                    </td>

                    <td>
                        <a href="http://exac.broadinstitute.org/about">Website</a>
                    </td>

                     <td>
                        <a href="ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.2/ExAC.r0.2.sites.vep.vcf.gz">File</a>

                    </td>


                    <td>
                        <a href="?vcf=http://s3.amazonaws.com/vcf.files/ExAC.r0.2.sites.vep.vcf.gz&species=Human&build=GRCh37">Show in vcf.iobio</a>
                    </td>
                </tr>


                <tr >
                    <td>
                        1000 Genomes Project 
                    </td>

                    <td>
                        <a href="http://www.1000genomes.org/data">Website</a>
                    </td>

                    <td>
                        <a href="ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/release/20130502/ALL.wgs.phase3_shapeit2_mvncall_integrated_v5.20130502.sites.vcf.gz">File</a>
                    </td>

                    <td>
                        <a href="?vcf=http://s3.amazonaws.com/vcf.files/ALL.wgs.phase3_shapeit2_mvncall_integrated_v5.20130502.sites.vcf.gz&species=Human&build=GRCh37">Show in vcf.iobio</a>
                    </td>
                </tr>

                <tr >
                    <td>
                        Illumina Platinum - NA12878
                    </td>

                    <td>
                        <a href="http://www.illumina.com/platinumgenomes/">Website</a>
                    </td>

                    <td>
                        <a href="ftp://ussd-ftp.illumina.com/IlluminaPlatinumGenomes_v7.0/merged_platinum/NA12877.vcf.gz">File</a>
                    </td>

                    <td>
                        <a href="?vcf=http://s3.amazonaws.com/iobio/variants/NA12878.autosome.PASS.vcf.gz&species=Human&build=GRCh37">Show in vcf.iobio</a>
                    </td>
                </tr>



                </table>

                <div style="margin-top: 40px"><strong>For sequence alignment files check out </strong><a style="margin-left: 37px" href="http://bam.iobio.io">bam.iobio</a></div>    
            </div>

            <div id="footer">                        
                <div class="logos">     
                    <div style="display:inline;font-size:50px;font-family:'Overlock SC', cursive;">Marthlab</div>     
                    <img src="assets/images/ustar-ucgd-logo.jpg" style="height:60px;"/>            
                    <a href="http://www.genetics.utah.edu/"><img src="assets/images/genetics_mainlogo3_lrg.png" style="height:50px"/></a>          
                </div>                      
            </div>

        </div>

        <div  id="showData" style="padding-top:5px">

            <div id="page" >
                <div>
                    <a href="javascript:void(0)" class="hide" id="report-problem" data-toggle="modal" data-target="#modal-report-problem">Report Problem</a>
                </div>
                <section id="banner" style="clear: both;">
    				
                    <div id="banner-title" >
                        <header><a href="/">vcf<span style="color:rgb(200,200,200)">.iobio.io</span></a></header>
                         <span id="vcf_file"></span>
                    </div>


                    <div id="banner-middle">
                       
                        <div id="vcf_selection">
                            <div id="current-build"></div>

                            <div style="text-align:center">
                                <div id="reference_selected"></div>
                                <div id="region_selected"></div>
                            </div>
                            <div id="samples-filter-header" class="" style="text-align:center">    <div id="label">
                                    <a id="show-sample-dialog"
                                       href="javascript:void(0)"
                                       class="btn hide"
                                       onclick="showSamplesDialog()"
                                       data-toggle="modal" 
                                       data-target="#sample-picker-modal">
                                       Filter samples
                                    </a>
                                </div>

                                <div id="sample-names"></div>
                            </div>

                        </div>
                    </div>



                    <div id="total-reads" >
                        <div >
                            <span id="value"> 0</span>
                        </div>
                        <div style="padding: 4px" >
                            <div id="number">&nbsp;</div>
                            <div id="label">variants sampled</div>
                        </div>
                        <div style="padding-top: 10px; ">
                            <img id="sample-more-button"   
                                 title="Sample More" 
                                 onClick="increaseSampling()"
                                 src="assets/images/more_sampling.png"
                                 style="width: 20px">
                             </img>
                        </div>

                    </div>
    				
    			</section>

                <section id="top" >

                   
                    <div id="chromosome-picker" class="panel">
                        <div class="title">References
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="ref_help" title="Select the reference sequence (e.g. chromosome) to view"></span>
                        </div>
                        <div class="samplingLoader big">Initializing<img src="assets/images//loading_dots.gif"/></div>
                        <div id="primary-references">
                        </div>    
                        <div id="other-references">
                             <label>Other references</label>
                             <select id="other-references-dropdown">
                               <option value=""></option>
                             </select>
                        </div>                    
                    </div>

                    <div id="variant-density-panel" class="panel">
                        <div class="title">Variant Density
                          <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="density_help" title="Distribution of variants across genome/region"></span>
                        </div>
                        <span class="svg-alt hint">(drag bottom chart to select region)</span>
                        <div class="samplingLoader big">Initializing<img src="assets/images//loading_dots.gif"/></div>
                        <div id="variant-density">
                        </div>    
                        <div id="variant-density-vf" >

                        </div>
                        <div id="variant-density-ref-vf">

                        </div>



                    </div>

                    <!-- Variant density modal -->
                    <div id="densityModal" class="modal fade" role="dialog">
                      <div class="modal-dialog" style="width:700px;">
                        <div class="modal-content">
                          <div class="modal-header">
                            <button type="button" class="close" data-dismiss="modal">&times;</button>
                            <h4 class="modal-title">Variant density</h4>
                          </div>
                          <div class="modal-body" style="overflow:auto;">
                            <div>
                              The variant density chart shows the distribution of variants across the genome. To zoom in on a specific reference sequence (chromosome), you can select the coloured sequence boxes under the distribution. The appearance of this chart is dependent on a number of different factors, but specifically, the type of sequencing undertaken, e.g. whole genome (WGS) or whole exome (WES).
                            </div>
                            <br>
                            <div><strong><i>Whole genome sequencing</i></strong></div>
                            <div>
                              For a WGS VCF file, the variant distribution should be evenly distributed across the whole genome, as shown for a human sample below.
                            </div>
                            <img title="WGS variant density" style="width:100%; padding-top:10px; padding-bottom:10px;"  src="images/density_wgs.png"></img>
                            <div>
                              You will notice that there are some areas of zero variant density in the above chart. Chromosomes 1, 3, 16, 19, and 20 are <i>metacentric</i> chromosomes, and have a centromeric region where no variants will be present, close to the centre of the chromosome. This gap is particularly large in chromosomes 1 and 16. Chromosome 9 is <i>submetacentric</i> and has a large centromere, but the centromere is offset leading to p and q arms of different lengths. Finally, chromosomes 13, 14, 15, 21, and 22 are <i>acrocentric</i>. Here the p arm is so short, it is difficult to observe, leading to the centromere appearing at the beginning of the chromosomes.
                            </div>
                            <br>
                            <div>
                              Observing gaps in coverage for these chromosomes is expected, but any other gaps could be indicative of problems with the variant calls. Male samples in the VCF file have reduced coverage on the X chromosome (since they only have one copy), so it is also expected that the average variant density is lower on this chromosome. Often variant calls are not generated for the Y chromosome, or pseudoautosomal and heterochromatic regions are removed, leading to few calls. 
                            </div>
                            <br>
                            <div><strong><i>Whole exome sequencing</i></strong></div>
                            <div>
                              For a WES VCF file containing many samples, the variant distribution will appear much more uneven, since the VCF file will only contain variants in exonic regions. Some of the gaps described for the WGS can still be observed, but there is still generally coverage across the whole chart. Any large gaps would be suggestive of data problems.
                            </div>
                            <img title="WES variant density" style="width:100%; padding-top:10px; padding-bottom:10px;"  src="images/density_wes.png"></img>
                          </div>
                          <div class="modal-footer">
                            <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                          </div>
                        </div>
                      </div>
                    </div>

                </section>
              
    			<section id="middle"  >

                        <div id="tstv-ratio" class="panel">
                            <div class="title">Ts/Tv Ratio
                              <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="tstv_help" title="Expect a value of approximately 2"></span>
                            </div>
                            <div class="samplingLoader">Sampling <img src="assets/images/loading_dots.gif"/></div>
                            <div id="ratio-panel" class="svg-alt">
                            </div>  
                            <div id="ratio-value" class="svg-alt" ></div>
                            <div class='no-values hide'>No values present</div>
                        </div>

                        <!-- Ts/Tv modal -->
                        <div id="tstvModal" class="modal fade" role="dialog">
                          <div class="modal-dialog" style="width:700px;">
                            <div class="modal-content">
                              <div class="modal-header">
                                <button type="button" class="close" data-dismiss="modal">&times;</button>
                                <h4 class="modal-title">Transition/transversion (Ts/Tv) ratio</h4>
                              </div>
                              <div class="modal-body" style="overflow:auto;">
                                <div>
                                  The transition / transversion (or Ts / Tv) ratio is calculated by looking at SNPs in the selected file, and is expected to have a value of the order of 2. The nucleotides Adenine (A) and Guanine (G) are purines, and Cytosine (C) and Thymine (T) are pyrimidines.
                                </div>
                                <div class="row vertical-align">
                                    <div class="col-xs-6">
                                      <img title="Transitions" style="width:100%;"  src="images/transitions.png"></img>
                                    </div>
                                    <div class="col-xs-6">
                                      A transition occurs when a purine mutates into another purine. There are four possible transition mutations, as shown in the figure to the left. 
                                    </div>
                                </div>
  
                                <div class="row vertical-align">
                                    <div class="col-xs-6">
                                      <img title="Transversions" style="width:100%;" src="images/transversions.png"></img>
                                    </div>
                                    <div class="col-xs-6">
                                      A transversion occurs when a pyrimidine mutates into another pyrimidine, and there are eight possible ways for this to occur.
                                    </div>
                                </div>
                                <div>
                                  If single base-pair mutations were purely random events, we would expect to see twice as many transversions as transitions, as there are twice as many ways for a transversion to occur. We would then expect to see a Ts/Tv ratio of 0.5. However, transitions are biologically more favourable, consequently for human DNA, we actually expect to see far more transitions than we would expect by chance: typically we see a Ts/Tv ratio closer to 2. If the observed Ts/Tv is significantly lower than 2, then it is a potential signal of data problems.
                                </div>
                              </div>
                              <div class="modal-footer">
                                <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                              </div>
                            </div>
                          </div>
                        </div>

                        <div id="allele-freq" class="panel">
                            <div class="title">Allele Frequency Spectrum
                              <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="af_help" title="AF is dependent on the number of samples"></span>
                            </div>
                            <div class="samplingLoader">Sampling <img src="assets/images/loading_dots.gif"/></div>
                            <svg id="allele-freq-histogram"/>
                            <div class='no-values hide'>No values present</div>
                        </div>
                        <div id="mut-spectrum" class="panel">
                            <div class="title">Base Changes
                              <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="bases_help" title="Frequency of observed mutations"></span>
                            </div>
                            <div class="samplingLoader">Sampling <img src="assets/images/loading_dots.gif"/></div>
                            <div class='no-values hide'>No values present</div>
                        </div>
                </section>

                <section id="bottom"  >


                        <div id="var-type" class="panel">
                            <div class="title">Variant Types
                              <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="variant_help" title="Distribution of different observed variant types"></span>
                            </div>
                            <div class="samplingLoader">Sampling <img src="assets/images/loading_dots.gif"/></div>
                            <div class='no-values hide'>No values present</div>
                        </div>

                        <!-- Variant types modal -->
                        <div id="variantModal" class="modal fade" role="dialog">
                          <div class="modal-dialog" style="width:700px;">
                            <div class="modal-content">
                              <div class="modal-header">
                                <button type="button" class="close" data-dismiss="modal">&times;</button>
                                <h4 class="modal-title">Variant types</h4>
                              </div>
                              <div class="modal-body" style="overflow:auto;">
                                <div>
                                  This chart shows a distribution of the variant types present in the selected VCF file. SNPs are by far the most prevalent type of genetic mutation, so it is expected that these dominate the distribution.
                                </div>
                                <div class="row vertical-align">
                                    <div class="col-xs-4">
                                      <img title="WGS variant types" style="width:100%; padding-top:15px; padding-bottom:15px;"  src="images/types_wgs.png"></img>
                                    </div>
                                    <div class="col-xs-8">
                                       For whole genome sequencing data, SNPs dominate, but there is typically an appreciable number of insertion (Ins) and deletion (Del) mutations, as well as <i>Other</i> variant types. These other variants include anything that doesn't fall into the previous categories. For example, consider the case where the reference sequence is ACTG, and sometimes the A is mutated into a C, but whenever it is, the T is always deleted. Rather than these two mutations appearing separately in the VCF file, since they are only ever observed together, the reference will be ACTG, and the alternate will be CCG (the A mutated to C, and the T deleted). This mutation is a combination of a SNP and a deletion, and is listed as a <i>complex</i> variant type and will appear in the <i>Other</i> category.
                                    </div>
                                </div>

                                <div class="row vertical-align">
                                    <div class="col-xs-4">
                                      <img title="WES variant types" style="width:100%; padding-top:15px; padding-bottom:15px;" src="images/types_wes.png"></img>
                                    </div>
                                    <div class="col-xs-8">
                                      It is much less likely to find structural variants, when looking at whole exome data. Large events are often not detected since the breakpoints fall outside of the exome sequencing regions. Also, insertions in, or deletions of large coding regions are likely to be more deleterious than in the non-coding regions captured in whole genome sequencing, and so there are less present in these regions. For these reasons, SNPs are expected to dominate the variant types distribution even more in exome sequencing.
                                    </div>
                                </div>
                                <div>
                                  This chart is primarily used to check that all expected variants are present. If you are expecting to see a complete VCF including insertions and deleted, but the file has been filtered to leave only SNPs, this chart will quickly help identify this.
                                </div>
                              </div>
                              <div class="modal-footer">
                                <button type="button" class="btn btn-default" data-dismiss="modal">Close</button>
                              </div>
                            </div>
                          </div>
                        </div>

                        <div id="indel-length" class="panel">
                            <div>
                                <div class="title" style="display:inline-block">Insertion &amp; Deletion Lengths
                                  <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="indel_help" title="Distribution of insertion and deletion lengths"></span>
                                </div>
                                <span class="checkbox" class="svg-alt" style="display:inline-block">
                                    <label>
                                        <input type="checkbox" checked value="true">Outliers
                                    </label>
                                </span>
                            </div>
                            <div class="samplingLoader">Sampling <img src="assets/images/loading_dots.gif"/></div>
   
                            <div id="indel-length-histogram"></div>

                            <div class="svg-alt">
                                <span style="font-weight:bold;font-size:13px;">Deletions</span>
                                <div id="legend-del"></div>
                                <span  style="font-weight:bold;font-size:13px;">Insertions</span>
                                <div id="legend-ins"></div>
                            </div>
                            
                            <div class='no-values hide'>No values present</div>
                        </div>

                        <div id="qual-distribution" class="panel">
                            <div class="title">Variant Quality
                              <span class="glyphicon glyphicon-info-sign" aria-hidden="true" id="quality_help" title="Distribution of the quality of observed variants"></span>
                            </div>
                            <div class="samplingLoader">Sampling <img src="assets/images/loading_dots.gif"/></div>
                            <svg id="qual-distribution-histogram"/>
                            <div class='no-values hide'>No values present</div>
                        </div>
                </section>

           
    		</div>
        </div>


<!-- Modal -->
<div class="modal fade" id="modal-report-problem" tabindex="-1" role="dialog">
  <div class="modal-dialog" role="document">
    <div class="modal-content">
        <div class="modal-header">
          <h2>Report Problem</h2>
          <a href="#close" class="btn-close" aria-hidden="true">×</a> <!--CHANGED TO "#close"-->
        </div>
        <div class="modal-body">
           <div>
             <span class="label">Your email</span><input id="report-problem-email" type="text" ></input>
          </div>
          <div style="margin-top:10px; margin-bottom:10px">
             <span class="label"</span>Description of problem<textarea id="report-problem-note" ></textarea>
          </div>
          <span class="label"</label>Screen Capture</span>
          <div id="report-problem-screen-capture">
        </div>    
        <div class="modal-footer">
          <a  id="report-problem-button" href="#email-report-problem"  class="btn">Report</a>  <!--CHANGED TO "#close"-->
          <a  class="btn" data-dismiss="modal">Cancel</a>
        </div>
    </div>
  </div>
</div>
</div>

<div class="modal fade"  id="sample-picker-modal" tabindex="-1" role="dialog">
  <div  class="modal-dialog" role="document">
    <div class="modal-content">
      <div class="modal-header" style="padding-top:10px;">
         <h4>Filter samples</h4>
          <a href="#close" class="btn-close" aria-hidden="true">×</a>         
      </div>
      <div class="modal-body">

        <div id="sample-picker" class="">
            <div class="vcf-sample loader hide" style="text-align:center">
              <span class="loader-label" >Initializing</span>
              <img src="assets/images/loading_dots.gif" >
            </div>
              
            <div id="vcf-sample-box" class="hide" style="text-align:center">
              <div id="vcf-sample-select-box"  style="display:inline-block;">
                <select id="vcf-sample-select"  data-placeholder="choose sample(s)...">
                </select>
              </div>
                
            </div>                            
        </div>

      </div>
      <div class="modal-footer" style="clear-both">
          <button type="button" id="sample-go-button" class="btn" data-dismiss="modal">Load</button>
      </div>
    </div>
  </div>
</div>
<!-- /Modal -->
        
        <script>
            var div = d3.select("body").append("div")   
                .attr("class", "tooltip")               
                .style("opacity", 0);
        </script>             

        <script type="text/javascript" src="app/vcf.iobio.js"></script>
        <script type="text/javascript" src="app/line.d3.js"></script>
        <script type="text/javascript" src="app/barChartAlt.d3.js"></script>
        <script type="text/javascript" src="app/histogram.d3.js"></script>
        <script type="text/javascript" src="app/groupedBar.d3.js"></script>
        <script type="text/javascript" src="app/genomeBuildHelper.js"></script>
        <script type="text/javascript" src="app/dataSelect.js"></script>
        <script type="text/javascript" src="app/app.js"></script>

	</body>
</html>
